velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -0,0 +1,221 @@
+          filters.PushFilter(
+              colIdx,
+              std::make_unique<::duckdb::ConstantFilter>(
+                  ::duckdb::ExpressionType::COMPARE_LESSTHANOREQUALTO,
+                  ::duckdb::Value(rangeFilter->upper())));
+        }
+      }
+    } break;
+
+    case common::FilterKind::kAlwaysFalse:
+    case common::FilterKind::kAlwaysTrue:
+    case common::FilterKind::kIsNull:
+    case common::FilterKind::kIsNotNull:
+    case common::FilterKind::kBoolValue:
+    case common::FilterKind::kBigintValuesUsingHashTable:
+    case common::FilterKind::kBigintValuesUsingBitmask:
+    case common::FilterKind::kDoubleRange:
+    case common::FilterKind::kFloatRange:
+    case common::FilterKind::kBytesRange:
+    case common::FilterKind::kBytesValues:
+    case common::FilterKind::kBigintMultiRange:
+    case common::FilterKind::kMultiRange:
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported filter in parquet reader: {}", filter->toString());
+  }
+}
+
+} // anonymous namespace
+
+ParquetRowReader::ParquetRowReader(
+    std::shared_ptr<::duckdb::ParquetReader> reader,
+    const dwio::common::RowReaderOptions& options,
+    memory::MemoryPool& pool)
+    : reader_(std::move(reader)), pool_(pool) {
+  auto& selector = *options.getSelector();
+  auto& filter = selector.getProjection();
+  rowType_ = selector.buildSelectedReordered();
+
+  std::vector<::duckdb::column_t> columnIds;
+  columnIds.reserve(rowType_->size());
+  duckdbRowType_.reserve(rowType_->size());
+  for (auto& node : filter) {
+    columnIds.push_back(node.column);
+    duckdbRowType_.push_back(reader_->return_types[node.column]);
+  }
+
+  std::vector<idx_t> groups;
+  for (idx_t i = 0; i < reader_->NumRowGroups(); i++) {
+    auto groupOffset = reader_->GetFileMetadata()->row_groups[i].file_offset;
+    if (groupOffset >= options.getOffset() &&
+        groupOffset < (options.getLength() + options.getOffset())) {
+      groups.push_back(i);
+    }
+  }
+
+  if (options.getScanSpec()) {
+    auto& scanSpec = *options.getScanSpec();
+    for (auto& colSpec : scanSpec.children()) {
+      VELOX_CHECK(
+          !colSpec->extractValues(),
+          "Subfield access is NYI in parquet reader");
+      if (colSpec->filter()) {
+        // TODO: remove linear search
+        uint64_t colIdx = std::find(
+                              reader_->names.begin(),
+                              reader_->names.end(),
+                              colSpec->fieldName()) -
+            reader_->names.begin();
+        VELOX_CHECK(
+            colIdx < reader_->names.size(),
+            "Unexpected columns name: {}",
+            colSpec->fieldName());
+        toDuckDbFilter(colIdx, colSpec->filter(), filters_);
+      }
+    }
+  }
+
+  reader_->InitializeScan(
+      state_, std::move(columnIds), std::move(groups), &filters_);
+}
+
+uint64_t ParquetRowReader::next(uint64_t size, velox::VectorPtr& result) {
```

Comment:
we dont seem to be using size, can we get rid of it?

Reply:
This comes from the basic interface. It's used in the other implementation. Also, this implementation should be improved to respect `size` param (but this would require a similar feature in DuckDB Parquet reader).